src/sentry/dynamic_sampling/models/transaction_adjustment_model.py

```diff
@@ -0,0 +1,190 @@
+    :param max_explicit_transactions: the maximum number of transactions that can have individual
+        rates set, the rest will have a common rate
+    :return: a tuple with the first element a mapping transaction-name->sampling-rate and the
+    second element the transaction rate for all other transactions (that are not in the dict)
+    """
+    # sort by transaction count
+    transactions = sorted(transactions, key=lambda x: x[1])
+    if len(transactions) <= max_explicit_transactions:
+        # we can get the ideal rate to all do a full resample
+        return AdjustedSampleRate(
+            explicit_rates=adjust_sample_rate_full(transactions, rate), global_rate=rate
+        )
+
+    # TODO I think we can find out which is the best option by looking at the distribution
+    #   if we compare the smallest rate with the max_explicit_rate_1 rate and the ratio is
+    #   around  1/rate then I think it is safe to go with resample_min otherwise resample_max
+    #   need to investigate and refine this idea.
+
+    # See what's gives better results, setting the rate of the smallest transactions or of
+    # the largest transactions (if we have just a few very large transactions is better to
+    # give them individual rates), if we have a few very small transactions we are better off
+    # specifying sample rate for the small transactions.
+    # The way we evaluate what's best is by maximizing the minimum number of samples returned
+    # by a transaction which is not sampled at 1.0
+    min_sample_size_x, explicit_rates_x, global_rate_x = adjust_sample_rate_max(
+        transactions, rate, max_explicit_transactions
+    )
+
+    min_sample_size_n, explicit_rates_n, global_rate_n = adjust_sample_rate_min(
+        transactions, rate, max_explicit_transactions
+    )
+
+    if min_sample_size_n < min_sample_size_x:
+        return AdjustedSampleRate(explicit_rates=explicit_rates_x, global_rate=global_rate_x)
+    else:
+        return AdjustedSampleRate(explicit_rates=explicit_rates_n, global_rate=global_rate_n)
+
+
+def adjust_sample_rate_full(
+    transactions: List[Tuple[str, int]], rate: float
+) -> MutableMapping[str, float]:
+    """
+    resample all transactions to their ideal size
+    """
+    transactions = copy(transactions)
+    ret_val = {}
+    num_transactions = total_transactions(transactions)
+    # calculate how many transactions we are allowed to keep overall
+    # this will allow us to pass transactions between different transaction types
+    total_budget = num_transactions * rate
+    while transactions:
+        num_types = len(transactions)
+        # We recalculate the budget per type every iteration to
+        # account for the cases where, in the previous step we couldn't
+        # spend all the allocated budget for that type.
+        budget_per_transaction_type = total_budget / num_types
+        name, count = transactions.pop(0)
+        if count < budget_per_transaction_type:
```

Comment:
Do we want to use the `<=` due to the equivalent behavior of the `else` in case `count == budget_per_transaction_type`?

Reply:
I find `<` more clear, if we don't have enough transactions i.e. `num_transactions < desired_transactions` then to this ... otherwise do that.
Of course `<=` also works and should give exact the same results ("minus any rounding errors from division).